govcd/user.go

```diff
@@ -0,0 +1,463 @@
+func (adminOrg *AdminOrg) GetUserByNameOrId(identifier string, willRefresh bool) (*OrgUser, error) {
+	if willRefresh {
+		err := adminOrg.Refresh()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for _, orgUser := range adminOrg.AdminOrg.Users.User {
+		if (orgUser.Name == identifier || orgUser.ID == identifier) && orgUser.Type == types.MimeAdminUser {
+
+			user := NewUser(adminOrg.client, adminOrg)
+
+			_, err := adminOrg.client.ExecuteRequest(orgUser.HREF, http.MethodGet,
+				types.MimeAdminUser, "error getting user: %s", nil, user.User)
+
+			return user, err
+		}
+	}
+	return nil, ErrorEntityNotFound
+}
+
+// GetRole finds a role within the organization
+func (adminOrg *AdminOrg) GetRole(roleName string) (*types.Reference, error) {
+
+	// There is no need to refresh the AdminOrg, until we implement CRUD for roles
+	for _, role := range adminOrg.AdminOrg.RoleReferences.RoleReference {
+		if role.Name == roleName {
+			return role, nil
+		}
+	}
+
+	return &types.Reference{}, ErrorEntityNotFound
+}
+
+// CreateUser creates an OrgUser from a full configuration structure
+// The timeOut variable is the maximum time we wait for the user to be ready
+// (This operation does not return a task)
+// Note that the timeout is not absolute, meaning that we only wait the
+// full amount of the timeout if the user has not been created until then.
+// This function returns as soon as the user has been created, which could be as
+// little as 50ms or as much as TimeOut - 50 ms
+// Mandatory fields are: Name, Role, Password.
+// https://code.vmware.com/apis/442/vcloud-director#/doc/doc/operations/POST-CreateUser.html
+func (adminOrg *AdminOrg) CreateUser(userConfiguration *types.User, timeOut time.Duration) (*OrgUser, error) {
+	err := validateUserForCreation(userConfiguration)
+	if err != nil {
+		return nil, err
+	}
+
+	userCreateHREF, err := url.ParseRequestURI(adminOrg.AdminOrg.HREF)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing admin org url: %s", err)
+	}
+	userCreateHREF.Path += "/users"
+
+	user := NewUser(adminOrg.client, adminOrg)
+
+	_, err = adminOrg.client.ExecuteRequest(userCreateHREF.String(), http.MethodPost,
+		types.MimeAdminUser, "error creating user: %s", userConfiguration, user.User)
+	if err != nil {
+		return nil, err
+	}
+
+	// Uncomment this line to show the user structure after creation
+	// ShowUser(*user.OrgUser)
+
+	// If there is a valid task, we try to follow through
+	if user.User.Tasks != nil && len(user.User.Tasks.Task) > 0 {
+		task := NewTask(adminOrg.client)
+		task.Task = user.User.Tasks.Task[0]
+		err = task.WaitTaskCompletion()
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Attempting to retrieve the user
+	// Since there is no task to wait for, we try to retrieve using a max of 200 attempts
+	// at intervals of 50 Ms
+	// Experience shows that it usually takes less than one second to get the result
+	var delayPerAttempt time.Duration = 50
```

Comment:
50ms sounds pretty often, have you checked - could you check if we even get replies that fast?

Reply:
50 ms is not a reply time. It's how long we wait before attempting the next retrieval

Reply:
OK, still feels too often/DOSy. What about 200ms as in vapp.go?

Reply:
Changed, and used Client.MaxRetryTimeout as suggested by Dainius, instead of a function parameter.

Reply:
Fantastic, thanks!